src/dataflow/src/sink/kafka.rs

```diff
@@ -176,105 +183,122 @@ pub fn kafka<G>(
-                            // retry sending this record at a later time.
-                            // Note that any other error will result in dropped
-                            // data as we will not attempt to resend it.
-                            // https://github.com/edenhill/librdkafka/blob/master/examples/producer.c#L188-L208
-                            // only retries on QueueFull so we will keep that
-                            // convention here.
-                            encoded_buffer = Some((encoded, count));
-                            activator.activate_after(Duration::from_secs(60));
+                        sink_metrics.rows_queued.set(queue.len() as u64);
+                        sink_metrics.messages_in_flight.set(in_flight as u64);
+                        if encoded_buffer.is_some() || !queue.is_empty() {
+                            // We need timely to reschedule this operator as we have pending
+                            // items that we need to send to Kafka
+                            activator.activate();
                             return true;
                         }
-                    } else {
-                        sink_metrics.messages_sent_counter.inc();
-                    }
 
-                    // Cache the Avro encoded data if we need to send again and
-                    // remember how many more times we need to send it
-                    if count > 1 {
-                        encoded_buffer = Some((encoded, count - 1));
-                    }
-                }
-
-                let in_flight = producer.in_flight_count();
-
-                sink_metrics.rows_queued.set(queue.len() as u64);
-                sink_metrics.messages_in_flight.set(in_flight as u64);
-                if encoded_buffer.is_some() || !queue.is_empty() {
-                    // We need timely to reschedule this operator as we have pending
-                    // items that we need to send to Kafka
-                    activator.activate();
-                    return true;
-                }
-
-                if in_flight > 0 {
-                    // We still have messages that need to be flushed out to Kafka
-                    // Let's make sure to keep the sink operator around until
-                    // we flush them out
-                    activator.activate_after(Duration::from_secs(5));
-                    return true;
-                }
+                        if in_flight > 0 {
+                            // We still have messages that need to be flushed out to Kafka
+                            // Let's make sure to keep the sink operator around until
+                            // we flush them out
+                            activator.activate_after(Duration::from_secs(5));
+                            return true;
+                        }
 
-                false
-            }
+                        false
+                    } else {
```

Comment:
The level of nesting in this function has made things pretty hard to follow. Probably time to extract into a helper function?

Reply:
I've tried to change it to check for is_none and return early in that case, which avoids the extra level of nesting. let me know if that feels good enough.

I'm happy to extract into a helper instead but it seems kind of hard because the innermost closure uses so many variables defined in outer scopes. If you feel strongly about this though please lmk

Reply:
I also changed it to bind the inner closure to a variable and then return the pair (closure, shutdownbutton) separately which also reduces a level of nesting